components/feature/tabs/src/main/java/mozilla/components/feature/tabs/tabstray/TabsTrayPresenter.kt

```diff
@@ -43,6 +45,12 @@ class TabsTrayPresenter(
 
     private suspend fun collect(flow: Flow<BrowserState>) {
         flow.map { it.toTabs(tabsFilter) }
+            .map { tabs ->
+                // Load the tab thumbnail from the memory or disk caches.
+                tabs.copy(list = tabs.list.map { tab ->
+                    tab.copy(thumbnail = thumbnailsUseCases.loadThumbnail(tab.id))
```

Comment:
This has a downside that I see after testing it in Sample Browser: since the in-memory version is null, we always load it from disk. This means the bitmap from the previous list is different from the one we just fetched now.

When the RecyclerView's DiffUtils checks if the tabs have changed to process a UI update, it sees that the references have changed so updates all the tab thumbnails even if the image itself hasn't changed.

Reply:
Override the `equals()` in the `Tab` data class to check for the `thumbnail.generationId` 
The generation ID changes whenever the bitmap is modified. This can be used as an efficient way to check if a bitmap has changed.(https://developer.android.com/reference/kotlin/android/graphics/Bitmap#getgenerationid)